elastic/eui
# Clean up copy/pasting of tabular content in EuiDataGrid and EuiBasic/InMemoryTable

## Summary

closes https://github.com/elastic/kibana/issues/177952

This PR adds a tabular content copy service which listens for copy events, detects hidden characters that both EuiDataGrid and EuiBasicTable now embed, and strips those characters and any undesired newlines/tabs and replaces them with only newlines/tabs that correspond with rows and cells.

- The inspiration/initial idea for this workaround came from [this StackOverflow response](https://stackoverflow.com/a/78151173/4294462).
  - I opted against using https://developer.mozilla.org/en-US/docs/Web/API/Clipboard/readText or `writeText()` because it only works on `https` and with permissions. I went with the more old school `clipboardData.setData()` instead: https://developer.mozilla.org/en-US/docs/Web/API/Element/copy_event

> [!warning]
> This will **not** work as expected for **virtualized data grids that scroll** (especially if the last cell is not visible and scrolled off the page). There is not a lot EUI can do about this as the cells are literally not rendered in the DOM - at most we could provide an `onCopy` API for consumers to hook into and provide their own cleaned data, but that would have to be a future feature request/enhancement. 

## QA

- Go to https://eui.elastic.co/pr_8019/storybook/index.html?path=/story/tabular-content-euidatagrid--playground
- Ctrl+A to select all or select a few rows
- [x] Open Google Spreadsheets and paste into it and confirm the grid row(s) paste in cleanly and as expected
- Go to https://eui.elastic.co/pr_8019#/tabular-content/tables#a-basic-table
- Copy 2 tables from anywhere on the page
- [x] Paste into Google Spreadsheets or a notepad type editor and confirm that the tables are stripped in terms of newlines but the other content is not

### General checklist

- Browser QA
    ~- [ ] Checked in both **light and dark** modes~
    ~- [ ] Checked in **mobile**~
- Docs site QA - N/A
- Code quality checklist
    ~- [ ] Updated **[visual regression tests](https://github.com/elastic/eui/blob/main/wiki/contributing-to-eui/testing/visual-regression-testing.md)**~
- Release checklist
    ~- [ ] If applicable, added the **breaking change** issue label (and filled out the breaking change checklist)~
- Designer checklist - N/A
